File: Java-Module/src/main/java/com/sim/javamodule/domain/product/domain/ProductRepository.java
package com.sim.javamodule.domain.product.domain;

import jakarta.persistence.LockModeType;
import org.jetbrains.annotations.NotNull;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Lock;
import org.springframework.data.jpa.repository.Query;
import org.springframework.transaction.annotation.Transactional;

import java.util.Optional;

public interface ProductRepository extends JpaRepository<Product, Long> {
    @Query("select p from Product p where p.id = :productId")
    @Lock(value = LockModeType.PESSIMISTIC_WRITE)
    Optional<Product> findProduct( Long productId);
}

File: Kotlin-Module/src/main/kotlin/com/sim/kotlin/domain/product/domain/ProductRepository.kt
package com.sim.kotlin.domain.product.domain

import jakarta.persistence.LockModeType
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Lock
import org.springframework.data.jpa.repository.Query

interface ProductRepository: JpaRepository<Product, Long> {

    @Query("select p from Product p where p.id = :productId")
    @Lock(value = LockModeType.PESSIMISTIC_WRITE)
    fun findProduct(productId: Long) : Product?
}